### ISSUE TITLE
Add async update api

### ISSUE THREAD

**Issue by felangel** [shorebirdtech/shorebird#123]
Something like this imagined Dart?
```dart
var updater = Updater();
var update = updater.checkForUpdate();
// Show user dialog about the update, including size, etc.
updater.beginDownload(update);
updater.statusCallback = function (updater) {
   if (updater.updateComplete) {
      // Show dialog offering restart?
   } 
};
```

**Comment by chimon2000** [comment]
If I wanted to push release notes with the update, how would those be conveyed?

**Comment by felangel** [comment]
Minor suggestions, lmk what you think.

```dart
final updater = Updater();

// Check whether there is an update available.
// Returns null if no update is available.
final update = await updater.checkForUpdate();

// Do nothing if there is no update.
if (update == null) return;

// Otherwise prompt the user whether they want to update.
final shouldUpdate = _showUpdatePrompt();

// Do nothing if the user declines to update.
if (!shouldUpdate) return;

// Otherwise, install the update.
try {
  await updater.installUpdate(
    update,
    onProgressChanged: (progress) {
      // update progress indicator
    },
  );
  // Show success dialog, prompt to restart.
} catch (error) {
  // Show failure dialog, prompt to retry.
}
```

**Comment by felangel** [comment]
> If I wanted to push release notes with the update, how would those be conveyed?

That would mostly likely be exposed from the Shorebird CLI when you create a patch.

**Comment by chimon2000** [comment]
> > If I wanted to push release notes with the update, how would those be conveyed?
> 
> That would mostly likely be exposed from the Shorebird CLI when you create a patch.

I'm guessing there would be some sort of success callback or another API call I could make to get that information.

**Comment by felangel** [comment]
> > > If I wanted to push release notes with the update, how would those be conveyed?
> > 
> > 
> > That would mostly likely be exposed from the Shorebird CLI when you create a patch.
> 
> I'm guessing there would be some sort of success callback or another API call I could make to get that information.

I think we could expose the release notes as part of the update object but totally open to suggestions.

**Comment by chimon2000** [comment]
> > > > If I wanted to push release notes with the update, how would those be conveyed?
> > > 
> > > 
> > > That would mostly likely be exposed from the Shorebird CLI when you create a patch.
> > 
> > I'm guessing there would be some sort of success callback or another API call I could make to get that information.
> 
> I think we could expose the release notes as part of the update object but totally open to suggestions.

Maybe something like


```dart
await updater.installUpdate(
    update,
    onProgressChanged: (progress) {
      // update progress indicator
    },
    onComplete: (release) {
       // release.notes from CLI
    }
  );
```

The one piece I'm not sure about is how these notes would be maintained after the user restarts the application but I am sure there's a similar pattern to what happens on desktop. I'm interested in how code push handles this.

**Comment by felangel** [comment]
> > > > > If I wanted to push release notes with the update, how would those be conveyed?
> > > > 
> > > > 
> > > > That would mostly likely be exposed from the Shorebird CLI when you create a patch.
> > > 
> > > 
> > > I'm guessing there would be some sort of success callback or another API call I could make to get that information.
> > 
> > 
> > I think we could expose the release notes as part of the update object but totally open to suggestions.
> 
> Maybe something like
> 
> ```dart
> await updater.installUpdate(
>     update,
>     onProgressChanged: (progress) {
>       // update progress indicator
>     },
>     onComplete: (release) {
>        // release.notes from CLI
>     }
>   );
> ```
> 
> The one piece I'm not sure about is how these notes would be maintained after the user restarts the application but I am sure there's a similar pattern to what happens on desktop. I'm interested in how code push handles this.

I think the release notes could be part of the initial update object:

```dart
final updater = Updater();

// Check whether there is an update available.
// Returns null if no update is available.
final update = await updater.checkForUpdate();

final releaseNotes = update?.releaseNotes;
```

**Comment by a14n** [comment]
This API suggest that the user has control on what is installed (via the `update` returned value). Is it intended?

For instance if we use checkForUpdate and get an updateWithPatch1 variable. Then a patch-2 is available. Will `installUpdate(updateWithPatch1)` install patch-1? or patch-2?

**Comment by eseidel** [comment]
> This API suggest that the user has control on what is installed (via the `update` returned value). Is it intended?
> 
> For instance if we use checkForUpdate and get an updateWithPatch1 variable. Then a patch-2 is available. Will `installUpdate(updateWithPatch1)` install patch-1? or patch-2?

I think it would need to install patch-1.

I see the user as having two actions/decisions here.
1. do I want to allow download of Xkb (e.g. spend money of my mobile bill)
2. do I want to restart now or later when I'm done with whatever I was in the middle of.

The reason I'd suggest patch-1, is that that was what they'd approved the download of.  I guess if patch-2 is close in size you could argue one should might as well do that. 🤷

**Comment by eseidel** [comment]
Actually I think it should be patch2 after further thought.  Installing patch1 (even if smaller) is a worse user experience.  That said, we just provide the APIs here, shorebird the service probably doesn't care which UX you choose, but our examples would like to help you get a great one for your users.

**Comment by eseidel** [comment]
I'm going to make this work:

```dart
final updater = Updater();

// Check whether there is an update available.
// Returns null if no update is available.
final update = await updater.checkForUpdate();

// Do nothing if there is no update.
if (update == null) return;

// Otherwise prompt the user whether they want to download update
final shouldDownload = _showUpdatePrompt(update);

// Do nothing if the user declines to update.
if (!shouldDownload) return;

// Otherwise, install the update.
try {
  await updater.installUpdate(update);
  // Show success dialog, prompt to restart.
} catch (error) {
  // Show failure dialog, prompt to retry.
}
```

I'm skipping progress indication (and release notes) for now.

**Comment by eseidel** [comment]
OK, so in terms of what to expose to the dev/user on Update?
```dart
class Update {
final String version;
final int bytes;
}
```

Or maybe that should be "Patch"?  And maybe patchNumber instead of version?  Or maybe still version, but it's the "version string" from the running app + the patch number?

**Comment by eseidel** [comment]
Going to do:
```dart
class Patch {
  final String baseVersion;
  final int number;
  final int bytes;
}
```

**Comment by chimon2000** [comment]
> > > > > > If I wanted to push release notes with the update, how would those be conveyed?
> > > > > 
> > > > > 
> > > > > That would mostly likely be exposed from the Shorebird CLI when you create a patch.
> > > > 
> > > > 
> > > > I'm guessing there would be some sort of success callback or another API call I could make to get that information.
> > > 
> > > 
> > > I think we could expose the release notes as part of the update object but totally open to suggestions.
> > 
> > 
> > Maybe something like
> > ```dart
> > await updater.installUpdate(
> >     update,
> >     onProgressChanged: (progress) {
> >       // update progress indicator
> >     },
> >     onComplete: (release) {
> >        // release.notes from CLI
> >     }
> >   );
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The one piece I'm not sure about is how these notes would be maintained after the user restarts the application but I am sure there's a similar pattern to what happens on desktop. I'm interested in how code push handles this.
> 
> I think the release notes could be part of the initial update object:
> 
> ```dart
> final updater = Updater();
> 
> // Check whether there is an update available.
> // Returns null if no update is available.
> final update = await updater.checkForUpdate();
> 
> final releaseNotes = update?.releaseNotes;
> ```

Yeah, I think that makes sense. Dug into code push docs for Microsoft to see how this works. Their CLI allows you to add a [description](https://learn.microsoft.com/en-us/appcenter/distribution/codepush/cli?source=recommendations#description-parameter) when pushing a release.

```bash
appcenter codepush release ... --description <descriptionOfChanges>
``` 

The description is then accessible from the client SDK using [`getCurrentPackage`](https://learn.microsoft.com/en-us/appcenter/distribution/codepush/rn-api-ref?source=recommendations#codepushgetcurrentpackage) or `getUpdateMetadata`, the former of which also includes whether or not the update previously ran.

```javascript
codePush.getCurrentPackage()
.then((update) => {
    // If the current app "session" represents the first time
    // this update has run, and it had a description provided
    // with it upon release, let's show it to the end user
    if (update.isFirstRun && update.description) {
        // Display a "what's new?" modal
    }
});
```

**Comment by eseidel** [comment]
I think this is done as part of @bryanoltman's recent work?

We don't track progress yet, but we do at least expose an async api.

**Comment by bryanoltman** [comment]
> I think this is done as part of @bryanoltman's recent work?
> 
> We don't track progress yet, but we do at least expose an async api.

Yep, we don't support progress at the moment, but this is otherwise done.

**Comment by eseidel** [comment]
Progress (and retries) keep coming up for users, especially on iOS.

**Comment by luatvudinh** [comment]
> > I think this is done as part of @bryanoltman's recent work?
> > We don't track progress yet, but we do at least expose an async api.
> 
> Yep, we don't support progress at the moment, but this is otherwise done.

The progress is needed in my project. Please let me know if you have any plan for it.

**Comment by adibone** [comment]
> > > I think this is done as part of @bryanoltman's recent work?
> > > We don't track progress yet, but we do at least expose an async api.
> > 
> > 
> > Yep, we don't support progress at the moment, but this is otherwise done.
> 
> The progress is needed in my project. Please let me know if you have any plan for it.

I don't know this will help or not, I just doing this kind of things to check the update.

`Future<void> checkForUpdates() async {
    final isUpdateAvailable =
        await shorebirdCodePush.isNewPatchAvailableForDownload();
    final isUpdateDownloaded =
        await shorebirdCodePush.isNewPatchReadyToInstall();

    if (isUpdateDownloaded) {
      if (mounted) {
        showConfirmDialog(
          context,
          title: 'There is new update available',
          subtitle: 'Update require restart app, do you want to continue?',
          onConfirm: () => Restart.restartApp(),
        );
      }
    } else if (isUpdateAvailable) {
      await shorebirdCodePush.downloadUpdateIfAvailable();
      checkForUpdates();
    }
  }`

**Comment by JycTheDev** [comment]
Having the release note is also a great idea 👌

**Comment by eseidel** [comment]
We've started re-writing our Dart API.  Early WIP posted, feedback welcome: https://github.com/shorebirdtech/updater/pull/225/files

**Comment by maheshj01** [comment]
This would be a great addition to more precisely deliver the updates.

 Heres a short demo I am assuming a safe update duration of 15 secs for this simuulation

https://github.com/user-attachments/assets/86cdd5a4-0e3d-4731-9a37-49feb55f6cb9

**Comment by bdero** [comment]
Async API shipped back in 2023 (per the comments above). The download-progress piece split off into #3525, #3593, and the broader v2 design at #3684. Closing this in favor of those.
